File: app/protocol.py
"""The wire contract, mirrored in `web/src/lib/protocol.ts`.

Change one, change the other: `tests/test_protocol_contract.py` parses both and
fails when they drift.
"""

from __future__ import annotations

import re
from typing import Literal

from pydantic import BaseModel, ConfigDict, field_validator

ROOM_CODE_LENGTH = 5
# No O/0/I/1: misread when the code is read out loud.
ROOM_CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
ROOM_CODE_PATTERN = re.compile(f"^[{ROOM_CODE_ALPHABET}]{{{ROOM_CODE_LENGTH}}}$")

PARTICIPANT_NAME_MAX = 24

SCORE_DELTA_MAX = 1000
SCORE_REASON_MAX = 80
LABEL_MAX = 60

# These four are the whole defence against a lying clock, so the browser has to
# agree with the server about them; the contract test holds both to them.
BUZZ_COMPENSATION_CAP_MS = 750
BUZZ_MIN_SAMPLES = 3
BUZZ_SAMPLE_MAX_AGE_MS = 60_000
BUZZ_SAMPLE_WINDOW = 15

TIMER_MAX_MS = 6 * 60 * 60 * 1000
TIMER_ADD_MS = 30_000

PUZZLE_TITLE_MAX = 60
PUZZLE_WORD_MAX = 40
# Two pairs is the smallest pool that can be got wrong. Sixteen is the ceiling
# on a phone: thirty-two words is already a list that has to be scrolled, and
# the pool is capped, not the evening — a long round is several puzzles.
PUZZLE_MIN_PAIRS = 2
PUZZLE_MAX_PAIRS = 16
PUZZLE_MAX_PER_ROOM = 20

Role = Literal["gm", "player"]
TimerState = Literal["idle", "running", "paused", "expired"]
PuzzleStatus = Literal["draft", "live", "closed"]

SOCKET_ERRORS = {
    "no_token": "no_token",
    "bad_token": "bad_token",
    "room_gone": "room_gone",
    "kicked": "kicked",
}

EVENT_ERRORS = {
    "forbidden": "forbidden",
    "bad_payload": "bad_payload",
    "unknown_participant": "unknown_participant",
    "nothing_to_undo": "nothing_to_undo",
    "locked": "locked",
    "no_round": "no_round",
    "unknown_puzzle": "unknown_puzzle",
    # Sent to a puzzle that is not taking answers, or an edit to one already sent.
    "puzzle_locked": "puzzle_locked",
    "too_many_puzzles": "too_many_puzzles",
}

# Listed so the contract test can hold both languages to the same strings: a
# typo in one is otherwise silent, and shows up as a button that does nothing.
CLIENT_EVENTS = frozenset(
    {
        "clock:ping",
        "buzz:press",
        "room:setScoresVisible",
        "score:adjust",
        "score:undo",
        "buzz:arm",
        "buzz:setLocked",
        "buzz:reset",
        "timer:set",
        "timer:start",
        "timer:pause",
        "timer:resume",
        "timer:reset",
        "timer:addTime",
        "puzzle:create",
        "puzzle:update",
        "puzzle:delete",
        "puzzle:send",
        "puzzle:close",
        "puzzle:submit",
    }
)

SERVER_EVENTS = frozenset(
    {
        "state:sync",
        "room:update",
        "participants:update",
        "scores:update",
        "clock:pong",
        "buzz:armed",
        "buzz:result",
        "buzz:cleared",
        "timer:update",
        "timer:expired",
        "puzzle:list",
        "puzzle:board",
        "puzzle:cleared",
        "puzzle:review",
        "error",
    }
)


class Payload(BaseModel):
    model_config = ConfigDict(extra="forbid")


def normalise_room_code(value: str) -> str:
    code = value.strip().upper()
    if not ROOM_CODE_PATTERN.match(code):
        raise ValueError("invalid room code")
    return code


def normalise_name(value: str) -> str:
    name = value.strip()
    if not name:
        raise ValueError("name required")
    if len(name) > PARTICIPANT_NAME_MAX:
        raise ValueError("name too long")
    return name


def normalise_label(value: str | None) -> str | None:
    if value is None:
        return None
    label = value.strip()
    if not label:
        return None
    if len(label) > LABEL_MAX:
        raise ValueError("label too long")
    return label


class JoinRequest(Payload):
    code: str
    name: str

    @field_validator("code")
    @classmethod
    def _code(cls, value: str) -> str:
        return normalise_room_code(value)

    @field_validator("name")
    @classmethod
    def _name(cls, value: str) -> str:
        return normalise_name(value)


class GmLoginRequest(Payload):
    password: str
    # Omitted opens a new room; supplied resumes an existing one.
    code: str | None = None

    @field_validator("code")
    @classmethod
    def _code(cls, value: str | None) -> str | None:
        return None if value is None else normalise_room_code(value)


class ClockPing(Payload):
    clientSentAt: int  # noqa: N815 - the wire format is camelCase, like the browser


class SetScoresVisible(Payload):
    """False stops sending the numbers, not merely hides them."""

    visible: bool


class ScoreAdjust(Payload):
    participantId: str  # noqa: N815
    delta: int
    reason: str | None = None

    @field_validator("delta")
    @classmethod
    def _delta(cls, value: int) -> int:
        if value == 0:
            raise ValueError("delta must be non-zero")
        if abs(value) > SCORE_DELTA_MAX:
            raise ValueError("delta out of range")
        return value

    @field_validator("reason")
    @classmethod
    def _reason(cls, value: str | None) -> str | None:
        if value is None:
            return None
        reason = value.strip()
        if not reason:
            return None
        if len(reason) > SCORE_REASON_MAX:
            raise ValueError("reason too long")
        return reason


class ScoreUndo(Payload):
    participantId: str  # noqa: N815


class NoPayload(Payload):
    """Pure verbs still go through the authorise-then-validate wrapper, so no GM
    event name is a hole where an unvalidated payload gets through."""


class BuzzPress(Payload):
    """Every field is client-controlled and none is trusted; see `compensate()`."""

    roundId: str  # noqa: N815
    clientSentAt: int  # noqa: N815
    clockOffsetMs: int  # noqa: N815
    clockDelayMs: int  # noqa: N815
    sampleCount: int  # noqa: N815
    # Shown to the GM, never used for ranking.
    jitterMs: int  # noqa: N815


class BuzzArm(Payload):
    label: str | None = None
    locked: bool = True

    @field_validator("label")
    @classmethod
    def _label(cls, value: str | None) -> str | None:
        return normalise_label(value)


class BuzzSetLocked(Payload):
    roundId: str  # noqa: N815
    locked: bool


class TimerSet(Payload):
    durationMs: int  # noqa: N815
    label: str | None = None

    @field_validator("durationMs")
    @classmethod
    def _duration(cls, value: int) -> int:
        if value <= 0 or value > TIMER_MAX_MS:
            raise ValueError("duration out of range")
        return value

    @field_validator("label")
    @classmethod
    def _label(cls, value: str | None) -> str | None:
        return normalise_label(value)


class TimerAddTime(Payload):
    deltaMs: int  # noqa: N815

    @field_validator("deltaMs")
    @classmethod
    def _delta(cls, value: int) -> int:
        if value == 0 or abs(value) > TIMER_MAX_MS:
            raise ValueError("delta out of range")
        return value


class ParticipantView(Payload):
    id: str
    name: str
    # None while scores are hidden: withheld on the wire, not just in the UI.
    score: int | None
    connected: bool
    joinedAt: int  # noqa: N815


class ScoreView(Payload):
    participantId: str  # noqa: N815
    score: int | None


class RoomView(Payload):
    code: str
    status: Literal["open", "closed"]
    scoresVisible: bool  # noqa: N815


class MeView(Payload):
    role: Role
    participantId: str | None = None  # noqa: N815
    name: str | None = None


class BuzzRoundView(Payload):
    roundId: str  # noqa: N815
    label: str | None
    locked: bool
    armedAt: int  # noqa: N815


class BuzzPressView(Payload):
    """`deltaMs` is the gap that decided the round, `arrivalDeltaMs` the gap the
    network would have produced. Both relative to the winner, whose row reads 0."""

    participantId: str  # noqa: N815
    rank: int
    deltaMs: int  # noqa: N815
    arrivalDeltaMs: int  # noqa: N815
    compensationMs: int  # noqa: N815
    delayMs: int  # noqa: N815
    jitterMs: int  # noqa: N815
    # False: the guardrails rejected the offset and this ranked by arrival.
    compensated: bool
    clamped: bool


class BuzzResult(Payload):
    roundId: str  # noqa: N815
    presses: list[BuzzPressView]


class TimerView(Payload):
    state: TimerState
    label: str | None
    durationMs: int | None  # noqa: N815
    # Absolute server time, set only while running.
    endsAt: int | None  # noqa: N815
    remainingMs: int  # noqa: N815
    serverNow: int  # noqa: N815


def normalise_title(value: str) -> str:
    title = value.strip()
    if not title:
        raise ValueError("title required")
    if len(title) > PUZZLE_TITLE_MAX:
        raise ValueError("title too long")
    return title


def normalise_word(value: str) -> str:
    word = " ".join(value.split())
    if not word:
        raise ValueError("word required")
    if len(word) > PUZZLE_WORD_MAX:
        raise ValueError("word too long")
    return word


class PuzzlePairInput(Payload):
    """Two words that belong together. `first` and `second` are a writing order
    and nothing more: the player is handed both loose, in one pool."""

    first: str
    second: str

    @field_validator("first", "second")
    @classmethod
    def _word(cls, value: str) -> str:
        return normalise_word(value)


def _validated_pairs(pairs: list[PuzzlePairInput]) -> list[PuzzlePairInput]:
    """Every word in the pool is checked against every other, not column against
    column: the words are all of one kind, and a word appearing twice would make
    two pairings equally right with only one of them scored."""
    if not PUZZLE_MIN_PAIRS <= len(pairs) <= PUZZLE_MAX_PAIRS:
        raise ValueError("wrong number of pairs")

    pool = [word.casefold() for pair in pairs for word in (pair.first, pair.second)]
    if len(set(pool)) != len(pool):
        raise ValueError("duplicate word")
    return pairs


class PuzzleCreate(Payload):
    title: str
    pairs: list[PuzzlePairInput]

    @field_validator("title")
    @classmethod
    def _title(cls, value: str) -> str:
        return normalise_title(value)

    @field_validator("pairs")
    @classmethod
    def _pairs(cls, value: list[PuzzlePairInput]) -> list[PuzzlePairInput]:
        return _validated_pairs(value)


class PuzzleUpdate(PuzzleCreate):
    puzzleId: str  # noqa: N815


class PuzzleId(Payload):
    """`puzzle:delete`, `puzzle:send` and `puzzle:close` all name one puzzle."""

    puzzleId: str  # noqa: N815


class PuzzleSubmit(Payload):
    """The left column is the server's own order, so an answer is just the right
    cards read top to bottom — one id per row, every card used exactly once."""

    puzzleId: str  # noqa: N815
    arrangement: list[str]


class PuzzlePairView(Payload):
    """The key. Sent to the game master, and to a player only once the puzzle is
    closed and there is nothing left to spoil."""

    first: str
    second: str


class PuzzleDraftView(Payload):
    id: str
    title: str
    status: PuzzleStatus
    position: int
    pairs: list[PuzzlePairView]
    submissionCount: int  # noqa: N815


class PuzzleSlotView(Payload):
    """One dealt word. `slotId` is opaque: the id of the slot holding this word
    for this participant, not the id of the pair it came from."""

    slotId: str  # noqa: N815
    word: str


class PuzzleMatchView(Payload):
    """Two slots that belong together. Unordered: which is which means nothing."""

    slotId: str  # noqa: N815
    partnerSlotId: str  # noqa: N815


class PuzzleBoardView(Payload):
    """What one participant's phone holds: the whole pool in `cards`, and the
    pairing so far in `arrangement` — every slot id, read two at a time, which
    is what the dragging rearranges."""

    puzzleId: str  # noqa: N815
    title: str
    status: PuzzleStatus
    cards: list[PuzzleSlotView]
    arrangement: list[str]
    submitted: bool
    submittedAt: int | None  # noqa: N815
    total: int
    # Both withheld until the puzzle closes: a score is a hint, and the key is
    # the whole answer.
    correct: int | None
    key: list[PuzzleMatchView] | None


class PuzzleAnswerView(Payload):
    """One pair the participant made. `expected` is what `first` belonged with,
    and is only worth reading when the pair is wrong."""

    first: str
    second: str
    expected: str
    correct: bool


class PuzzleSubmissionView(Payload):
    participantId: str  # noqa: N815
    name: str
    submittedAt: int  # noqa: N815
    correct: int
    total: int
    answers: list[PuzzleAnswerView]


class PuzzleReviewView(Payload):
    puzzleId: str  # noqa: N815
    title: str
    status: PuzzleStatus
    total: int
    pairs: list[PuzzlePairView]
    submissions: list[PuzzleSubmissionView]
    # Named so the GM knows whether to wait or to close the puzzle.
    pending: list[str]


class StateSync(Payload):
    room: RoomView
    participants: list[ParticipantView]
    me: MeView
    round: BuzzRoundView | None
    presses: list[BuzzPressView]
    timer: TimerView
    # GM only: the authored puzzles, and the results of the one being watched.
    puzzles: list[PuzzleDraftView]
    review: PuzzleReviewView | None
    # Player only: their own dealt board, if a puzzle is live.
    puzzle: PuzzleBoardView | None
    serverNow: int  # noqa: N815

File: web/src/lib/protocol.ts
/**
 * Mirror of `app/protocol.py`, which is the authority: the server re-validates
 * everything here. Change one, change the other — the contract test parses both.
 */

export const ROOM_CODE_LENGTH = 5;
/** No O/0/I/1: misread when the code is read out loud. */
export const ROOM_CODE_ALPHABET = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789";
export const ROOM_CODE_PATTERN = new RegExp(`^[${ROOM_CODE_ALPHABET}]{${ROOM_CODE_LENGTH}}$`);

export const PARTICIPANT_NAME_MAX = 24;

export const SCORE_DELTA_MAX = 1000;
export const SCORE_REASON_MAX = 80;
export const SCORE_QUICK_DELTAS = [1, -1, 5, -5] as const;
export const LABEL_MAX = 60;

/** These four must agree with the server; the contract test checks it. */
export const BUZZ_COMPENSATION_CAP_MS = 750;
export const BUZZ_MIN_SAMPLES = 3;
export const BUZZ_SAMPLE_MAX_AGE_MS = 60000;
export const BUZZ_SAMPLE_WINDOW = 15;

export const CLOCK_PING_INTERVAL_MS = 2000;
/** Fired on connect and when a round is armed: a device that just woke up must
 *  not buzz on a stale offset. */
export const CLOCK_PING_BURST = 5;
/** Not back to back: five identical round trips tell the minimum filter nothing
 *  the first did not, and ~200ms of burst makes a device compensable in time
 *  for a round armed right after it joined. */
export const CLOCK_PING_BURST_GAP_MS = 50;

export const TIMER_MAX_MS = 6 * 60 * 60 * 1000;
export const TIMER_ADD_MS = 30000;
export const PUZZLE_TITLE_MAX = 60;
export const PUZZLE_WORD_MAX = 40;
/** Two pairs is the smallest pool that can be got wrong; eight is twelve rows
 *  short of a scroll hunt on a phone, at sixteen words. */
export const PUZZLE_MIN_PAIRS = 2;
export const PUZZLE_MAX_PAIRS = 8;
export const PUZZLE_MAX_PER_ROOM = 20;

export const TIMER_PRESETS_MS = [15000, 30000, 60000, 120000, 300000] as const;
export const TIMER_WARN_MS = 10000;

export function normaliseRoomCode(value: string): string | null {
  const code = value.trim().toUpperCase();
  return ROOM_CODE_PATTERN.test(code) ? code : null;
}

export function normaliseName(value: string): string | null {
  const name = value.trim();
  if (!name || name.length > PARTICIPANT_NAME_MAX) return null;
  return name;
}

export function normaliseLabel(value: string): string | null {
  const label = value.trim();
  return label ? label.slice(0, LABEL_MAX) : null;
}

export type Role = "gm" | "player";
export type TimerState = "idle" | "running" | "paused" | "expired";
export type PuzzleStatus = "draft" | "live" | "closed";

export const SOCKET_ERRORS = {
  no_token: "no_token",
  bad_token: "bad_token",
  room_gone: "room_gone",
  kicked: "kicked",
} as const;

export type SocketErrorCode = keyof typeof SOCKET_ERRORS;

export const EVENT_ERRORS = {
  forbidden: "forbidden",
  bad_payload: "bad_payload",
  unknown_participant: "unknown_participant",
  nothing_to_undo: "nothing_to_undo",
  locked: "locked",
  no_round: "no_round",
  unknown_puzzle: "unknown_puzzle",
  /** Sent to a puzzle that is not taking answers, or an edit to one already sent. */
  puzzle_locked: "puzzle_locked",
  too_many_puzzles: "too_many_puzzles",
} as const;

export type EventErrorCode = keyof typeof EVENT_ERRORS;

export interface ParticipantView {
  id: string;
  name: string;
  /** null while scores are hidden: withheld on the wire, not just in the UI. */
  score: number | null;
  connected: boolean;
  joinedAt: number;
}

export interface ScoreView {
  participantId: string;
  score: number | null;
}

export interface RoomView {
  code: string;
  status: "open" | "closed";
  scoresVisible: boolean;
}

export interface MeView {
  role: Role;
  participantId?: string | null;
  name?: string | null;
}

export interface BuzzRoundView {
  roundId: string;
  label: string | null;
  locked: boolean;
  armedAt: number;
}

/**
 * `deltaMs` is the gap that decided the round, `arrivalDeltaMs` the gap the
 * network would have produced. Both relative to the winner.
 */
export interface BuzzPressView {
  participantId: string;
  rank: number;
  deltaMs: number;
  arrivalDeltaMs: number;
  compensationMs: number;
  delayMs: number;
  jitterMs: number;
  /** false: the guardrails rejected the offset and this ranked by arrival. */
  compensated: boolean;
  clamped: boolean;
}

export interface BuzzResult {
  roundId: string;
  presses: BuzzPressView[];
}

export interface TimerView {
  state: TimerState;
  label: string | null;
  durationMs: number | null;
  /** Absolute, in *server* time, and set only while running. */
  endsAt: number | null;
  remainingMs: number;
  serverNow: number;
}

/** The key. The game master always has it; a player only once the puzzle closed. */
export interface PuzzlePairView {
  first: string;
  second: string;
}

export interface PuzzleDraftView {
  id: string;
  title: string;
  status: PuzzleStatus;
  position: number;
  pairs: PuzzlePairView[];
  submissionCount: number;
}

/**
 * One dealt word. `slotId` is opaque: the id of the slot holding this word for
 * this participant, not of the pair it came from — which is what keeps the key
 * off the wire.
 */
export interface PuzzleSlotView {
  slotId: string;
  word: string;
}

/** Two slots that belong together. Unordered: which is which means nothing. */
export interface PuzzleMatchView {
  slotId: string;
  partnerSlotId: string;
}

/** `cards` is the whole pool; `arrangement` is the pairing so far — every slot
 *  id, read two at a time, which is what the dragging rearranges. */
export interface PuzzleBoardView {
  puzzleId: string;
  title: string;
  status: PuzzleStatus;
  cards: PuzzleSlotView[];
  arrangement: string[];
  submitted: boolean;
  submittedAt: number | null;
  total: number;
  /** Both null until the puzzle closes: a score is a hint, the key is the answer. */
  correct: number | null;
  key: PuzzleMatchView[] | null;
}

/** One pair the participant made. `expected` is what `first` belonged with, and
 *  is only worth reading when the pair is wrong. */
export interface PuzzleAnswerView {
  first: string;
  second: string;
  expected: string;
  correct: boolean;
}

export interface PuzzleSubmissionView {
  participantId: string;
  name: string;
  submittedAt: number;
  correct: number;
  total: number;
  answers: PuzzleAnswerView[];
}

export interface PuzzleReviewView {
  puzzleId: string;
  title: string;
  status: PuzzleStatus;
  total: number;
  pairs: PuzzlePairView[];
  submissions: PuzzleSubmissionView[];
  /** Who has not answered yet, so the game master knows whether to wait. */
  pending: string[];
}

export interface StateSync {
  room: RoomView;
  participants: ParticipantView[];
  /** Who this socket is, so the UI can highlight its own row. */
  me: MeView;
  /** The open round and what has been pressed into it, so a reload mid-round
   *  comes back to the buzzer rather than to an empty lobby. */
  round: BuzzRoundView | null;
  presses: BuzzPressView[];
  timer: TimerView;
  /** Game master only: the authored puzzles, and the results being watched. */
  puzzles: PuzzleDraftView[];
  review: PuzzleReviewView | null;
  /** Player only: their own dealt board, when a puzzle is live. */
  puzzle: PuzzleBoardView | null;
  serverNow: number;
}

export interface ClockPing {
  clientSentAt: number;
}

export interface SetScoresVisible {
  visible: boolean;
}

export interface ScoreAdjust {
  participantId: string;
  delta: number;
  reason?: string | null;
}

export interface ScoreUndo {
  participantId: string;
}

/** Every field is client-controlled and none is trusted; the server bounds it. */
export interface BuzzPress {
  roundId: string;
  clientSentAt: number;
  clockOffsetMs: number;
  clockDelayMs: number;
  sampleCount: number;
  /** Shown to the GM, never used for ranking. */
  jitterMs: number;
}

export interface BuzzArm {
  label?: string | null;
  locked?: boolean;
}

export interface BuzzSetLocked {
  roundId: string;
  locked: boolean;
}

export interface TimerSet {
  durationMs: number;
  label?: string | null;
}

export interface TimerAddTime {
  deltaMs: number;
}

/** `first` and `second` are a writing order and nothing more: the player is
 *  handed both loose, in one pool. */
export interface PuzzlePairInput {
  first: string;
  second: string;
}

export interface PuzzleCreate {
  title: string;
  pairs: PuzzlePairInput[];
}

export interface PuzzleUpdate {
  puzzleId: string;
  title: string;
  pairs: PuzzlePairInput[];
}

/** `puzzle:delete`, `puzzle:send` and `puzzle:close` all name one puzzle. */
export interface PuzzleId {
  puzzleId: string;
}

/** The pool in the order the player has put it in, read two at a time: every
 *  card used exactly once, which is the only arrangement the server accepts. */
export interface PuzzleSubmit {
  puzzleId: string;
  arrangement: string[];
}

export interface ClientToServerEvents {
  "clock:ping": (payload: ClockPing) => void;
  "buzz:press": (payload: BuzzPress) => void;
  "room:setScoresVisible": (payload: SetScoresVisible) => void;
  "score:adjust": (payload: ScoreAdjust) => void;
  "score:undo": (payload: ScoreUndo) => void;
  "buzz:arm": (payload: BuzzArm) => void;
  "buzz:setLocked": (payload: BuzzSetLocked) => void;
  "buzz:reset": () => void;
  "timer:set": (payload: TimerSet) => void;
  "timer:start": () => void;
  "timer:pause": () => void;
  "timer:resume": () => void;
  "timer:reset": () => void;
  "timer:addTime": (payload: TimerAddTime) => void;
  "puzzle:create": (payload: PuzzleCreate) => void;
  "puzzle:update": (payload: PuzzleUpdate) => void;
  "puzzle:delete": (payload: PuzzleId) => void;
  "puzzle:send": (payload: PuzzleId) => void;
  "puzzle:close": (payload: PuzzleId) => void;
  "puzzle:submit": (payload: PuzzleSubmit) => void;
}

export interface ServerToClientEvents {
  "state:sync": (payload: StateSync) => void;
  "room:update": (payload: RoomView) => void;
  "participants:update": (payload: ParticipantView[]) => void;
  "scores:update": (payload: ScoreView[]) => void;
  "clock:pong": (payload: { clientSentAt: number; serverNow: number }) => void;
  "buzz:armed": (payload: BuzzRoundView) => void;
  "buzz:result": (payload: BuzzResult) => void;
  "buzz:cleared": (payload: { roundId: string }) => void;
  "timer:update": (payload: TimerView) => void;
  "timer:expired": (payload: { label: string | null }) => void;
  "puzzle:list": (payload: PuzzleDraftView[]) => void;
  "puzzle:board": (payload: PuzzleBoardView) => void;
  "puzzle:cleared": (payload: { puzzleId: string }) => void;
  "puzzle:review": (payload: PuzzleReviewView) => void;
  error: (payload: { code: EventErrorCode | string; message: string }) => void;
}
